os/binfmt/binfmt_loadbinary.c

```diff
@@ -91,37 +91,67 @@ int load_binary(int binary_idx, FAR const char *filename, load_attr_t *load_attr
 	struct tcb_s *tcb;
 #endif
 
-	/* Allocate the load information */
-
-	bin = (FAR struct binary_s *)kmm_zalloc(sizeof(struct binary_s));
-	if (!bin) {
-		berr("ERROR: Failed to allocate binary_s\n");
-		errcode = ENOMEM;
-		goto errout_with_bin;
-	}
-
-	/* Initialize the binary structure */
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD
+	bin = load_attr->binp;
+	/* If we find a non-null value for bin, it means that
+	 * we are in a reload scenario.
+	 */
+	if (bin) {
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
+			berr("ERROR: Failed to find copy of data section from previous load\n");
+			goto errout_with_bin;
+		}
+
+		memcpy(bin->datastart, bin->data_backup, bin->datasize);
+		memset(bin->bssstart, 0, bin->bsssize);
+		bin->reload = false;
+	} else {
+#endif
 
-	bin->filename = filename;
-	bin->exports = NULL;
-	bin->nexports = 0;
-	bin->filelen = load_attr->bin_size;
-	bin->offset = load_attr->offset;
-	bin->stacksize = load_attr->stack_size;
-	bin->priority = load_attr->priority;
-	bin->compression_type = load_attr->compression_type;
+		/* Allocate the load information */
+
+		bin = (FAR struct binary_s *)kmm_zalloc(sizeof(struct binary_s));
+		if (!bin) {
+			berr("ERROR: Failed to allocate binary_s\n");
+			errcode = ENOMEM;
+			goto errout_with_bin;
+		}
+
+		/* Initialize the binary structure */
+
+		bin->filename = filename;
+		bin->exports = NULL;
+		bin->nexports = 0;
+		bin->filelen = load_attr->bin_size;
+		bin->offset = load_attr->offset;
+		bin->stacksize = load_attr->stack_size;
+		bin->priority = load_attr->priority;
+		bin->compression_type = load_attr->compression_type;
 #ifdef CONFIG_APP_BINARY_SEPARATION
-	bin->ramsize = load_attr->ram_size;
+		bin->ramsize = load_attr->ram_size;
 #endif
 
-	/* Load the module into memory */
+		/* Load the module into memory */
 
-	ret = load_module(bin);
-	if (ret < 0) {
-		errcode = -ret;
-		berr("ERROR: Failed to load program '%s': %d\n", filename, errcode);
-		goto errout_with_bin;
+		ret = load_module(bin);
+		if (ret < 0) {
+			errcode = -ret;
+			berr("ERROR: Failed to load program '%s': %d\n", filename, errcode);
+			goto errout_with_bin;
+		}
+
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD
+		bin->data_backup = kmm_malloc(bin->datasize);
```

Comment:
For data section backup, how about allocating after RO section in user heap? Because we cannot estimate the each app's data size, so it is hard to fix the kernel heap size before build.

Reply:
It is possible that user heap might get corrupted, which might cause to reload the application. So, if data_section kept in user heap gets corrupted, then there will be crash or wrong operation after reload. There is no way to confirm the integrity of the user heap.

Reply:
As per our discussion, I have moved the data section backup into the ro region.